vmdk_plugin/shared_plugin/main.go

```diff
@@ -0,0 +1,111 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+// A Shared Docker Data Volume plugin - main
+
+import (
+	"flag"
+	"fmt"
+	"os"
+	"os/signal"
+	"reflect"
+	"runtime"
+	"syscall"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/docker/go-plugins-helpers/volume"
+	"github.com/vmware/docker-volume-vsphere/vmdk_plugin/drivers/shared"
+	"github.com/vmware/docker-volume-vsphere/vmdk_plugin/utils/config"
+	"github.com/vmware/docker-volume-vsphere/vmdk_plugin/utils/log_formatter"
+	"github.com/vmware/docker-volume-vsphere/vmdk_plugin/utils/plugin_server"
+)
+
+// main for docker-volume-vsphere
+// Parses flags, initializes and mounts refcounters and finally initializes the server.
+func main() {
+	var driver volume.Driver
+
+	// Get options from ENV (where available), and from command line.
+	// ENV takes precedence, so we can modify it in Docker plugin install
+	logEnv := os.Getenv("VDVS_LOG_LEVEL")
+	logLevel := &logEnv
+	if *logLevel == "" {
+		logLevel = flag.String("log_level", "", "Logging Level")
+	}
+	configFile := flag.String("config", config.DefaultSharedPluginConfigPath, "Configuration file path")
+	driverName := flag.String("driver", "", "Volume driver")
+
+	flag.Parse()
+
+	log_formatter.LogInit(logLevel, nil, configFile)
+
+	// Load the configuration if one was provided.
+	c, err := config.Load(*configFile)
+	if err != nil {
+		log.Warningf("Failed to load config file %s: %v", *configFile, err)
+	}
+
+	// If no driver provided on the command line, use the one in the
+	// config file or the default.
+	if *driverName == "" {
+		if err == nil && c.Driver != "" {
+			*driverName = c.Driver
+		} else {
+			*driverName = sharedDriver
+		}
+	}
+
+	// The windows plugin only supports the vsphere driver.
+	if runtime.GOOS == "windows" && *driverName != sharedDriver {
+		msg := fmt.Sprintf("Shared Plugin only supports the %s driver on Windows, ignoring parameter driver = %s.",
+			sharedDriver, c.Driver)
+		log.Warning(msg)
+		fmt.Println(msg)
+		*driverName = sharedDriver
+	}
+
+	log.WithFields(log.Fields{
+		"driver":    *driverName,
+		"log_level": *logLevel,
+		"config":    *configFile,
+	}).Info("Starting plugin ")
+
+	// The Shared driver doesn't depend on the config file for options
+	if *driverName == sharedDriver {
+		driver = shared.NewVolumeDriver(plugin_server.MountRoot, *driverName)
+	} else {
+		log.Warning("Unknown driver or invalid/missing driver options, exiting - ", *driverName)
+		os.Exit(1)
+	}
+
+	if reflect.ValueOf(driver).IsNil() == true {
+		log.Warning("Error in driver initialization exiting - ", *driverName)
+		os.Exit(1)
+	}
+
+	server := plugin_server.NewPluginServer(*driverName, &driver)
```

Comment:
So, this will instantiate a vmdk or shared or windows plugin server? It should instantiate based on the selection in plugin_server.go.